core/src/main/java/org/infinispan/commands/write/RemoveExpiredCommand.java

```diff
@@ -0,0 +1,179 @@
+package org.infinispan.commands.write;
+
+import org.infinispan.commands.CommandInvocationId;
+import org.infinispan.commands.Visitor;
+import org.infinispan.commons.equivalence.Equivalence;
+import org.infinispan.commons.util.CollectionFactory;
+import org.infinispan.commons.util.Util;
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.ExpiryHelper;
+import org.infinispan.container.entries.MVCCEntry;
+import org.infinispan.context.Flag;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.lifecycle.ComponentStatus;
+import org.infinispan.metadata.Metadata;
+import org.infinispan.notifications.cachelistener.CacheNotifier;
+import org.infinispan.util.TimeService;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.Objects;
+import java.util.Set;
+
+
+/**
+ * Removes an entry that is expired from memory
+ *
+ * @author William Burns
+ * @since 8.0
+ */
+public class RemoveExpiredCommand extends RemoveCommand {
+   public static final int COMMAND_ID = 58;
+   private static final Log log = LogFactory.getLog(RemoveExpiredCommand.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   protected Long lifespan;
+   protected TimeService timeService;
+
+   public RemoveExpiredCommand() {
+      // The value matcher will always be the same, so we don't need to serialize it like we do for the other commands
+      this.valueMatcher = ValueMatcher.MATCH_EXPECTED_OR_NULL;
+   }
+
+   public RemoveExpiredCommand(Object key, Object value, Long lifespan, CacheNotifier notifier,
+           Equivalence valueEquivalence, TimeService timeService, CommandInvocationId commandInvocationId) {
+      //valueEquivalence can be null because this command never compares values.
+      super(key, value, notifier, null, valueEquivalence, commandInvocationId);
+      this.lifespan = lifespan;
+      this.timeService = timeService;
+      this.valueMatcher = ValueMatcher.MATCH_EXPECTED_OR_NULL;
+   }
+
+   public void init(CacheNotifier notifier, Configuration configuration, TimeService timeService) {
+      super.init(notifier, configuration);
+      this.timeService = timeService;
+   }
+
+   /**
+    * Performs an expiration on a specified entry
+    *
+    * @param ctx invocation context
+    * @return null
+    */
+   @Override
+   public Object perform(InvocationContext ctx) throws Throwable {
+      MVCCEntry e = (MVCCEntry) ctx.lookupEntry(key);
+      if (e != null && !e.isRemoved()) {
+         Object value = e.getValue();
+         // If the provided lifespan is null, that means it is a store removal command, so we can't compare lifespan
+         if (lifespan == null) {
+            if (valueMatcher.matches(e, value, e.getValue(), valueEquivalence)) {
```

Comment:
It does do the expiration holding onto the locks.  The problem is this is invoked without holding the lock first, which means the value could change between when it saw it expired and before it finally obtains the lock.  The value and lifespan check is a last ditch check to make sure we don't remove a value that has since been added if possible.